segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -547,366 +678,538 @@ private void validateSystemSnapshot(SystemSnapshotRecord systemSnapshot) {
                 }
             }
         }
+        return CompletableFuture.completedFuture(null);
     }
 
     /**
      * Read contents from file.
      */
-    private byte[] getContents(String snapshotFile) throws Exception {
-        Exception e = null;
-        for (int i = 0; i < config.getMaxJournalReadAttempts(); i++) {
-            try {
-                val info = chunkStorage.getInfo(snapshotFile).get();
-                val h = ChunkHandle.readHandle(snapshotFile);
-                byte[] contents = new byte[Math.toIntExact(info.getLength())];
-                long fromOffset = 0;
-                int remaining = contents.length;
-                while (remaining > 0) {
-                    int bytesRead = chunkStorage.read(h, fromOffset, remaining, contents, Math.toIntExact(fromOffset)).get();
-                    remaining -= bytesRead;
-                    fromOffset += bytesRead;
-                }
-                return contents;
-            } catch (Exception ex) {
-                e = ex;
-                log.warn("SystemJournal[{}] Error while reading journal {}.", containerId, snapshotFile, e);
-            }
-        }
-        if (null != e) {
-            throw e;
-        }
-        return new byte[0];
+    private CompletableFuture<byte[]> getContents(String snapshotFile) {
+        val isReadDone = new AtomicBoolean();
+        val attempt = new AtomicInteger();
+        val lastException = new AtomicReference<Throwable>();
+        val retValue = new AtomicReference<byte[]>();
+        // Try config.getMaxJournalReadAttempts() times.
+        return Futures.loop(
+                () -> attempt.get() < config.getMaxJournalReadAttempts() && !isReadDone.get(),
+                () -> chunkStorage.getInfo(snapshotFile)
+                        .thenComposeAsync(info -> {
+                            val h = ChunkHandle.readHandle(snapshotFile);
+                            // Allocate buffer to read into.
+                            retValue.set(new byte[Math.toIntExact(info.getLength())]);
+
+                            val fromOffset = new AtomicLong();
+                            val remaining = new AtomicInteger(retValue.get().length);
+
+                            // Continue until there is still data remaining to be read.
+                            return Futures.loop(
+                                        () -> remaining.get() > 0,
+                                        () -> chunkStorage.read(h, fromOffset.get(), remaining.get(), retValue.get(), Math.toIntExact(fromOffset.get())),
+                                        bytesRead -> {
+                                            remaining.addAndGet(-bytesRead);
```

Comment:
What if you're unable to make progress? if bytesRead is 0, then you should be logging or at least asserting (so it bubbles up in tests).

Reply:
check added